weasyl/commishinfo.py

```diff
@@ -20,39 +52,259 @@ def convert_currency(target):
+    """
+    valuecode = _charmap_to_currency_code(valuecode)
+    targetcode = _charmap_to_currency_code(targetcode)
+    rates = _fetch_rates()
+    if not rates:
+        # in the unlikely event of an error, invalidate our rates
+        # (to try fetching again) and return value unaltered
+        _fetch_rates.invalidate()
+        return 1.0
+    rates = rates["rates"]
+    rates["USD"] = 1.0
+    r1 = rates.get(valuecode)
+    r2 = rates.get(targetcode)
+    if not (r1 and r2):
+        raise WeasylError("Unexpected")
+    return r2 / r1
 
-    content = d.execute("SELECT content FROM commishdesc WHERE userid = %i", [userid], ["element"])
 
+def select_list(userid):
+    query = d.engine.execute("SELECT classid, title, amount_min, amount_max, settings, priceid FROM commishprice"
+                             " WHERE userid = %(userid)s ORDER BY classid, title", userid=userid)
+    classes = d.engine.execute("SELECT classid, title FROM commishclass WHERE userid = %(id)s ORDER BY title", id=userid)
+    content = d.engine.execute("SELECT content FROM commishdesc WHERE userid = %(id)s", id=userid).scalar()
+    tags = d.engine.execute("SELECT DISTINCT tag.title, sma.settings FROM searchtag tag "
+                            "join searchmapartist sma on sma.tagid = tag.tagid "
+                            "join login l on l.userid = sma.targetid "
+                            "where l.userid = %(userid)s", userid=userid)
+    tags_extract = [{"t": i.title, "s": i.settings} for i in tags]
     return {
+        "userid": userid,
         "class": [{
-            "classid": i[0],
-            "title": i[1],
-        } for i in d.execute("SELECT classid, title FROM commishclass WHERE userid = %i ORDER BY title", [userid])],
+            "classid": i.classid,
+            "title": i.title,
+        } for i in classes],
         "price": [{
-            "classid": i[0],
-            "title": i[1],
-            "amount_min": i[2],
-            "amount_max": i[3],
-            "settings": i[4],
-            "priceid": i[5],
-        } for i in query if "a" not in i[4]] + [{
-            "classid": i[0],
-            "title": i[1],
-            "amount_min": i[2],
-            "amount_max": i[3],
-            "settings": i[4],
-            "priceid": i[5],
-        } for i in query if "a" in i[4]],
+            "classid": i.classid,
+            "title": i.title,
+            "amount_min": i.amount_min,
+            "amount_max": i.amount_max,
+            "settings": i.settings,
+            "priceid": i.priceid,
+        } for i in query if "a" not in i.settings] + [{
+            "classid": i.classid,
+            "title": i.title,
+            "amount_min": i.amount_min,
+            "amount_max": i.amount_max,
+            "settings": i.settings,
+            "priceid": i.priceid,
+        } for i in query if "a" in i.settings],
         "content": content if content else "",
+        "tags": [i["t"] for i in tags_extract if 'n' not in i["s"]],
```

Comment:
Are these the only places `tags_extract` is used? If so, you should be able to write `.fetchall()` on the query and `[i.title for i in tags if 'n' not in i.settings]` here, which should be easier to read.